runtime/addrmap.c

```diff
@@ -0,0 +1,134 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                 Stephen Dolan, University of Cambridge                 */
+/*                                                                        */
+/*   Copyright 2015 University of Cambridge                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include "caml/config.h"
+#include "caml/memory.h"
+#include "caml/addrmap.h"
+
+#define MAX_CHAIN 100
+
+static uintnat pos_initial(struct addrmap* t, value key)
+{
+  uintnat pos = (uintnat)key;
+  pos *= 0xcc9e2d51;
+  pos ^= (pos >> 17);
```

Comment:
Where does this hash function comes from?  The "multiply by 0xcc9e2d51" is also in Murmurhash 3, but not the "shift and xor".  It would be nice to give a reference.